### PR TITLE
feat: role binding backfill

### PR DESCRIPTION
**Pull request by james-milligan**
## This PR

- backfill the flagd-kubernetes-sync role binding on upgrade

### Related Issues

https://github.com/open-feature/open-feature-operator/issues/292

### Follow-up Tasks

This PR expands one of the options for resolving the bug described in the issue above. On startup a grouroutine is started with a 5 second timeout (to allow cache to be ready), and all pods with the `openfeature.dev/enabled` annotation set to `"true"` will have their service accounts added to the `flagd-kubernetes-sync` role binding, restoring its state to that which it was prior to the upgrade.

### How to test

Upgrade helm and verify that the flagd connection is reestablished.

### CODE REVIEW

File: this.yaml [hunk 1]
```diff
@@ -0,0 +1,41 @@
+apiVersion: core.openfeature.dev/v1alpha2
```
Comment (skyerus) [inline this.yaml:1]: should this be included?

File: config/manager/kustomization.yaml [hunk 2]
```diff
@@ -12,5 +12,5 @@ apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
 images:
 - name: controller
-  newName: controller
-  newTag: latest
+  newName: ghcr.io/open-feature/open-feature-operator
```
[the remark this answers is not in the document]
Comment (james-milligan) [reply config/manager/kustomization.yaml]: resolved

File: help.yaml [hunk 3]
```diff
@@ -0,0 +1,23 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
[the remark this answers is not in the document]
Comment (james-milligan) [reply help.yaml]: resolved

File: this.yaml (hunk 1 above)
Comment (james-milligan) [reply this.yaml]: resolved

File: webhooks/pod_webhook.go [hunk 4]
```diff
@@ -54,6 +54,7 @@ type PodMutator struct {
 // BackfillPermissions recovers the state of the flagd-kubernetes-sync role binding in the event of upgrade
 func (m *PodMutator) BackfillPermissions(ctx context.Context) {
 	for i := 0; i < 5; i++ {
+		fmt.Println("im running")
```
Comment (beeme1mr) [inline webhooks/pod_webhook.go]: 😄

Comment (james-milligan) [reply webhooks/pod_webhook.go]: you caught that before I removed it :(

File: webhooks/pod_webhook.go [hunk 5]
```diff
@@ -47,6 +51,38 @@ type PodMutator struct {
 	Log                       logr.Logger
 }
 
+// BackfillPermissions recovers the state of the flagd-kubernetes-sync role binding in the event of upgrade
+func (m *PodMutator) BackfillPermissions(ctx context.Context) {
+	for i := 0; i < 5; i++ {
+		if i == 5 {
```
Comment (Kavindu-Dodan) [inline webhooks/pod_webhook.go:57]: Let's move this out of the for loop. This gives more focus on the operations we attempt to perform

Plus, at this point our flagd deployments are unstable. Should we avoid operator start?

**Comment by toddbaert** [comment]
I could be missing something, but I still seem to have the issue after and upgrade. Here are the steps I took:

- built/pushed my docker image: `toddbaert/ofo:main`
- built charts: `make build && make helm-package`
- edited rendered.taml to use `toddbaert/ofo:main`
- deployed workload and verified everything worked

Then I applied the fix:

- checked out this branch
- edited `Makefile`, updating version to `v0.2.24`
- built/pushed my docker image: `toddbaert/ofo:fix`
- built charts: `make build && make helm-package`
- edited rendered.taml to use `toddbaert/ofo:fix`
- did `helm upgrade ofo .` from chart dir
- changed a flag value on the workload, saw:

```
"end-to-end\" is forbidden: User \"s │
│ ystem:serviceaccount:default:default\" cannot get resource \"featureflagconfigurations\" in API group \
```

**Comment by toddbaert** [comment]
@james-milligan 

As discussion on slack, the root cause of my issue was that I was using the old invalid annotation `openfeature.dev: "enabled"` not `openfeature.dev/enabled: "true"` (which works as expected).

As you noted, we may want to add support for the deprecated annotation too.

File: webhooks/pod_webhook.go (hunk 5 above)
Comment (james-milligan) [reply webhooks/pod_webhook.go]: This is a timeout allowing for the cache to be set up, which is managed in a separate groutine (unfortunately there is no 'ready' channel to hook into so the loop is required) -  it does not take into account the health of the deployments, and will backfill permissions even if the pod is crashing, i.e. 
```
NAME                             READY   STATUS             RESTARTS     AGE
busybox-curl-2-67d494d8c-4ldw8   1/2     CrashLoopBackOff   1 (2s ago)   8s
```
will still have its service account added as a subject to the role binding if its annotations are valid.

Ive cleaned up this loop, moving the timeout error to a drop out once the loop has finished

**Comment by james-milligan** [comment]
> @james-milligan
> 
> As discussed on on slack, the root cause of my issue was that I was using the old invalid annotation `openfeature.dev: "enabled"` not `openfeature.dev/enabled: "true"` (which works as expected).
> 
> As you noted, we may want to add support for the deprecated annotation too.
> 
> But otherwise, this works as expected. Nice job!

Ive added in support for the `core.openfeature.dev: "enabled"` annotation, as well as expanded the test to include it

File: webhooks/pod_webhook.go [hunk 6]
```diff
@@ -383,3 +415,21 @@ func setSecurityContext() *corev1.SecurityContext {
 		},
 	}
 }
+
+func OpenFeatureEnabledAnnotationIndex(o client.Object) []string {
+	val, ok := o.(*corev1.Pod).ObjectMeta.Annotations["openfeature.dev/enabled"]
```
[the remark this answers is not in the document]
Comment (james-milligan) [reply webhooks/pod_webhook.go]: you're right, ive resolved this and simplified the remaining statements

File: webhooks/pod_webhook.go [hunk 7]
```diff
@@ -47,6 +51,34 @@ type PodMutator struct {
...
 
+// BackfillPermissions recovers the state of the flagd-kubernetes-sync role binding in the event of upgrade
+func (m *PodMutator) BackfillPermissions(ctx context.Context) {
+	for i := 0; i < 5; i++ {
+		// fetch all pods with the "openfeature.dev/enabled" annotation set to "true"
+		podList := &corev1.PodList{}
+		err := m.Client.List(context.Background(), podList, client.MatchingFields{"metadata.annotations.openfeature.dev/enabled": "true"})
+		if err != nil {
+			if !goErr.Is(err, &cache.ErrCacheNotStarted{}) {
+				m.Log.Error(err, err.Error())
```
[the remark this answers is not in the document]
Comment (james-milligan) [reply webhooks/pod_webhook.go]: ive updated this error log to provide more context via the message argument

File: main.go [hunk 8]
```diff
@@ -178,7 +192,18 @@ func main() {
...
-	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
+	ctx := ctrl.SetupSignalHandler()
+	errChan := make(chan error, 1)
+	go func(chan error) {
+		if err := mgr.Start(ctx); err != nil {
+			errChan <- err
+		}
+	}(errChan)
+
+	setupLog.Info("starting backfill")
```
Comment (toddbaert) [inline main.go:203]: I think we can improve this message a bit.

Comment (james-milligan) [reply main.go]: updated to `restoring flagd-kubernetes-sync cluster role binding subjects from current cluster state`

**Review by toddbaert — APPROVED** [review]

File: webhooks/pod_webhook.go [hunk 9]
```diff
@@ -47,6 +51,43 @@ type PodMutator struct {
 	Log                       logr.Logger
 }
 
+// BackfillPermissions recovers the state of the flagd-kubernetes-sync role binding in the event of upgrade
+func (m *PodMutator) BackfillPermissions(ctx context.Context) {
+	for i := 0; i < 5; i++ {
+		// fetch all pods with the "openfeature.dev/enabled" annotation set to "true"
+		podList := &corev1.PodList{}
+		err := m.Client.List(ctx, podList, client.MatchingFields{"metadata.annotations.openfeature.dev/enabled": "true"})
```
[the remark this answers is not in the document]
Comment (james-milligan) [reply webhooks/pod_webhook.go]: introduced constant

**Comment by james-milligan** [comment]
resolved test race condition through introducing a 'complete' channel to the `BackfillPermissions` method, this is not used in `main` as the process can happen asynchronously as the ` mgr.Start(ctx)` method is blocking

**Review by Kavindu-Dodan — APPROVED** [review]

**Review by skyerus — APPROVED** [review]

**Pull request merged**